gwpy/table/table.py

```diff
@@ -754,31 +895,51 @@ def cluster(self, index, rank, window):
+
+        To cluster an `EventTable` (``table``) with the omicron clustering
+        algorithm, with a `window` of `0.1`, and maximize over `snr`:
+
+        >>> table.cluster('omicron', 'snr', 0.1)
         """
-        if window <= 0.0:
-            raise ValueError('Window must be a positive value')
-
-        # Generate index and rank vectors that are ordered
-        orderidx = numpy.argsort(self[index])
-        col = self[index][orderidx]
-        param = self[rank][orderidx]
-
-        # Find all points where the index vector changes by less than window
-        # and divide the resulting array into clusters of adjacent points
-        clusterpoints = numpy.where(numpy.diff(col) <= window)[0]
-        sublists = numpy.split(clusterpoints,
-                               numpy.where(numpy.diff(clusterpoints) > 1)[0]+1)
-
-        # Add end-points to each cluster and find the index of the maximum
-        # point in each list
-        padded_sublists = [numpy.append(s, numpy.array([s[-1]+1]))
-                           for s in sublists]
-        maxidx = [s[numpy.argmax(param[s])] for s in padded_sublists]
-
-        # Construct a mask that removes all points within clusters and
-        # replaces them with the maximum point from each cluster
-        mask = numpy.ones_like(col, dtype=bool)
-        mask[numpy.concatenate(padded_sublists)] = False
+
+        # Perform the clustering, without rewriting the table, just
+        # by appending a cluster ID to each trigger
+        # a deepcopy is used here because of the 'tstart' and 'tend'
+        # columns are rewritten using the omicron option. Meaning
+        # that if the cluster() method is called an another time
+        # without reloading the EventTable, the Tstart and Tend
+        # are the updated ones, not the original ones for the trigger
+        # which represent a cluster.
+        triggers = copy.deepcopy(self.clustering(index, window))
+
+        # Add a trigger index
+        triggers['idx'] = numpy.arange(len(self))
+
+        # store column to maximise over
+        param = triggers[rank]
+
+        # Get index of the trigger with highest value in column rank
+        # for each cluster
+        # Also get tstart_min and tend_max
+        maxidx = []
+        tstart_list = []
+        tend_list = []
+        for idx in triggers.group_by('clusterID').groups.keys:
+            mask = triggers['clusterID'] == idx[0]
+            maxidx.append(triggers[mask][numpy.argmax(param[mask])]['idx'])
+            if index == 'omicron':
+                tstart_list.append(numpy.min(triggers[mask]['tstart']))
+                tend_list.append(numpy.max(triggers[mask]['tend']))
+
+        # Apply these values to tile representing the cluster, only for
+        # the omicron method. Might need to use it as well for other method
+        if index == 'omicron':
+            triggers['tstart'][maxidx] = tstart_list
+            triggers['tend'][maxidx] = tend_list
```

Comment:
Please add the necessary logic to manipulate all other columns that are modified by the C++ version of this method, including `flow/fhigh` etc. We should strive to have the python method reproduce the C++ method exactly (including all parameter values).

Reply:
I included all of your recommandations. 

Appart from codeclimate and cognitive complexity, it looks fine. I'll let you check codeclimate outputs to see if it really matters.